src/fifelse.c

```diff
@@ -201,105 +201,104 @@ SEXP fifelseR(SEXP l, SEXP a, SEXP b, SEXP na) {
+        } else {
+          error(_("Argument #%d is of type %s, however argument #2 is of type %s. "
+                  "Please make sure all output values have the same type."),
+                  i*2+2, type2char(TYPEOF(thens)), type2char(type0));
+        }
       }
-      if (!R_compute_identical(PROTECT(getAttrib(value0,R_ClassSymbol)),  PROTECT(getAttrib(outs,R_ClassSymbol)), 0)) {
-        error(_("Argument #%d has different class than argument #2, "
-                "Please make sure all output values have the same class."), i*2+2);
+      if (!naout && !R_compute_identical(PROTECT(getAttrib(value0,R_ClassSymbol)),  PROTECT(getAttrib(thens,R_ClassSymbol)), 0)) {
+        if (idefault) {
+          error(_("Resulting value has different class than 'default'. "
+                  "Please make sure that both arguments have the same class."));
+        } else {
+          error(_("Argument #%d has different class than argument #2, "
+                  "Please make sure all output values have the same class."), i*2+2);
+        }
       }
-      UNPROTECT(2);
-      if (isFactor(value0)) {
-        if (!R_compute_identical(PROTECT(getAttrib(value0,R_LevelsSymbol)),  PROTECT(getAttrib(outs,R_LevelsSymbol)), 0)) {
-          error(_("Argument #2 and argument #%d are both factor but their levels are different."), i*2+2);
+      UNPROTECT(2); // class(value0), class(thens)
+      if (!naout && isFactor(value0)) {
+        if (!R_compute_identical(PROTECT(getAttrib(value0,R_LevelsSymbol)),  PROTECT(getAttrib(thens,R_LevelsSymbol)), 0)) {
+          if (idefault) {
+            error(_("Resulting value and 'default' are both type factor but their levels are different."));
+          } else {
+            error(_("Argument #2 and argument #%d are both factor but their levels are different."), i*2+2);
+          }
         }
-        UNPROTECT(2);
+        UNPROTECT(2); // levels(value0), levels(thens)
       }
     }
-    int64_t len1 = xlength(outs);
-    if (len1!=len0 && len1!=1) {
-      error(_("Length of output value #%d must either be 1 or length of logical condition."), i*2+2);
+    len1 = xlength(thens);
+    if (len1 != len0 && len1 != 1) {
+      if (idefault) {
+        error(_("Length of 'default' must be 1 or %"PRId64"."), len0);
+      } else {
+        error(_("Length of output value #%d must either be 1 or length of logical condition."), i*2+2);        
+      }
     }
-    int64_t amask = len1>1 ? INT64_MAX : 0;
-    const int *restrict pcons = LOGICAL(cons);
-    const bool imask = i==0;
-    int64_t l=0; // how many this case didn't satisfy; i.e. left for next case
-    switch(TYPEOF(outs)) {
+    int64_t amask = len1>1 ? INT64_MAX : 0, conmask = xlength(whens)>1 ? INT64_MAX : 0;
+    switch(TYPEOF(ans)) {
     case LGLSXP: {
-      const int *restrict pouts = LOGICAL(outs);
+      const int *restrict pthens;
+      if (!naout) pthens = LOGICAL(thens); // the content is not useful if out is NA_LOGICAL scalar
```

Comment:
(Applies to other cases too): Should the `if (naout)` branch be handled outside and before the switch statement?

Reply:
I had a similar thought about reorganizing the code... but probably better to explore as a follow-up.